A quickstart demo to showcase Hudi functionalities using docker along with support for integration-tests

bvaradar opened the pull request:
Comes with foundations for adding docker integration tests. Docker images built with Hadoop 2.8.4 Hive 2.3.3 and Spark 2.3.1

Demo using docker containers with  documentation

bvaradar commented:
@vinothchandar : The build is failing here 

---------------------
[ERROR] Failed to execute goal on project hoodie-hadoop-base-docker: Could not resolve dependencies for project com.uber.hoodie:hoodie-hadoop-base-docker:pom:0.4.4-SNAPSHOT: Could not transfer artifact com.uber.hoodie:hoodie-hadoop-docker:jar:0.4.4-SNAPSHOT from/to Maven repository (https://central.maven.org/maven2/): Host name 'central.maven.org' does not match the certificate subject provided by the peer (CN=repo1.maven.org, O="Sonatype, Inc", L=Fulton, ST=MD, C=US) -> [Help 1]
[ERROR]
===============


This is a brand new module. If I run "mvn clean package", it passes but "mvn clean install", it fails. Have you noticed this before when new modules were added to hudi ?

n3nash commented on docs/quickstart.md:
```diff
@@ -293,6 +293,453 @@ hive>
+
+Stock Tracker data will be used to showcase both different Hudi Views and the effects of Compaction. 
+
+Take a look at the director `docker/demo/data`. There are 2 batches of stock data - each at 1 minute granularity. 
```
directory*

n3nash commented on docs/quickstart.md:
```diff
@@ -293,6 +293,453 @@ hive>
+cat docker/demo/data/batch_1.json | kafkacat -b kafkabroker -t stock_ticks -P 
+
+To check if the new topic shows up, use
+kafkacat -b kafkabroker -L -J | jq .
```
Let's add a sample output here to ensure what people should compare against..

bvaradar replied:
Fixed.

sungjuly commented on docs/quickstart.md:
````diff
@@ -14,11 +14,11 @@ Check out code and pull it into Intellij as a normal maven project.
 Normally build the maven project, from command line
 ```
-$ mvn clean install -DskipTests
+$ mvn clean install -DskipTests -skipITs
````
`-skipITs` should be `-DskipITs`

bvaradar commented on docs/quickstart.md:
````diff
@@ -293,6 +293,453 @@ hive>
+partitions, there is no need to run hive-sync
+
+```
+cat /Users/varadarb/Uber/sync/varadarb.dev.uber.com/home/uber/code/hudi_ws/docker/demo/data/batch_2.json | kafkacat -b kafkabroker -t stock_ticks -P 
````
Fixed.

bvaradar commented on docs/quickstart.md:
````diff
@@ -293,6 +293,453 @@ hive>
+Again, You can use Hudi CLI to manually schedule and run compaction
+
+```
+^[[Aroot@adhoc-1:/opt#   /var/hoodie/ws/hoodie-cli/hoodie-cli.sh
````
Done.

bvaradar commented on docs/quickstart.md:
```diff
@@ -293,6 +293,453 @@ hive>
+# Schedule a compaction. This will use Spark Launcher to schedule compaction
+hoodie:stock_ticks->compaction schedule
+....
+Compaction instance : 20180910234509
```
Thanks @sungjuly . I have rebased and this issue should be fixed now.

bvaradar commented:
@sungjuly @n3nash :  Thanks a lot for providing the feedback. I have revamped this PR and made few improvements. Here is the list:

1. Reimplement docker files to reduce image sizes. I am seeing 3-4x improvements in docker sizes (from 4 - 5 GB to 0.8-1.5GB). 
2. I have published all the docker images to docker-hub (temporarily using my docker-hub account) so that there is no need to build the docker images as part of running the script. 
3. Made both docker images and compose follow convention so that a single image can satisfy both use-cases
   (a) Standalone - Docker images that are published to Docker-hub have inbuilt hudi jars so that they are self-contained.
  (b) Mounting Convention - The images allow for overriding those hudi jars with locally built jars allowing for the same setup to be used for running integ-testing against code-changes locally (look at the compose script and hoodie-integ/pom.xml)
4.  Updates quickstart docs to reflect the changes above

I have marked this PR as WIP since I need to showcase  incremental view as part of demo. 

@n3nash @sungjuly : Can you please follow the quickstart steps and let me know your experience. 
Our aim is to have an easy-to-use and hassle-free hudi quickstart. I would appreciate your feedback on that.

sungjuly commented on docs/quickstart.md:
````diff
@@ -293,6 +293,511 @@ hive>
+
+```
+cd docker
+./setup_demo.sh
````
I've encountered this. It seems it's related docker version. Here's my env to figure out problems. 

* docker-engine: 18.06.1-ce (there's no way to downgrade 17.12 since the official site only support latest version for Mac)
* docker-compose: 1.22.0

```
➜  docker git:(docker) ✗ ./setup_demo.sh
Removing network compose_default
Creating network "compose_default" with the default driver
Creating kafkabroker ...
Creating zookeeper                 ... error
Creating hive-metastore-postgresql ...
Creating namenode                  ...
Creating kafkabroker               ... error
ERROR: for zookeeper  Cannot create container for service zookeeper: Conflict. The container name "/zookeeper" is already in use by container "8fb174b9eb8c40bdc20e0ed1b042a793fe49ec3dcea1f353c250b482c7c80019". You have to remove (or rename) that container to be able to reuse that name.

ERROR: for kafkabroker  Cannot create container for service kafka: Conflict. The container name "/kafkabroker" is alreCreating hive-metastore-postgresql ... error
ame) that container to be able to reuse that name.
```

bvaradar replied:
@sungjuly : ok, can you check "docker container ls" and see if there are stopped/running containers with the same name. If you find them, can you remove the containers using "docker rm". 

This is an one-time issue for those who tried this PR before. I have moved the compose script to a different file but the container names are same. Removing the docker containers should hopefully fix the issue. 

For a different reason, I reverted my docker (mac) installation to "factory defaults" and that also fixed the problem.

sungjuly replied:
docker env was cleaned up all things before I tested. Here's output for `docker container ls`  

```
➜  docker git:(docker) ✗ docker container ls
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
```

bvaradar commented:
@sungjuly: I am also online in hoodie slack channel. Ping me there if you need quicker reply.

Thanks,
Balaji.V

sungjuly commented:
@bvaradar would you please share the link for hoodie slack channel? I don't see any information, thank you!

sungjuly commented:
nvmd, I found this - https://github.com/uber/hudi/issues/143#issuecomment-423682393

bvaradar commented:
Tests were failing because of log-size limit issue. Fixed as part of PR : https://github.com/uber/hudi/pull/465

n3nash commented on docker/compose/docker-compose_hadoop284_hive233_spark231.yml:
```diff
@@ -0,0 +1,215 @@
+services:
+
+  namenode:
+    image: varadarb/hudi-hadoop_2.8.4-namenode:latest
```
Is this correct ? Should the username be here ? @bvaradar

bvaradar replied:
@n3nash :
I am not aware of any headless docker hub account . I have created a ticket to provide more context and assigned to @vinothchandar  

https://github.com/uber/hudi/issues/469

Once the ticket is resolved, we can replace the image locations in Dockerfile, pom.xml and in compose scripts.

n3nash commented on docker/compose/docker-compose_hadoop284_hive233_spark231.yml:
```diff
@@ -0,0 +1,215 @@
+      - /tmp/hadoop_data:/hadoop/dfs/data
+
+  historyserver:
+    image: varadarb/hudi-hadoop_2.8.4-history:latest
```
same here

bvaradar commented on docs/quickstart.md:
````diff
@@ -293,6 +293,985 @@ hive>
+exit
+```
+
+#### Step 5 (b): Run Spark-SQl Queries
````
@sungjuly @vinothchandar @n3nash : added spark-sql example here

sungjuly replied:
👍 I've tested with new scenarios. It worked properly! Separately creating MOR/COW tables is a good idea to explain internal works. Thank you for your work as a user!

vinothchandar replied:
+1 It is so much better to test stuff now.. :)  Thank you for doing this, even as co-creator of the project :)  ha ha.

vinothchandar commented on docs/quickstart.md:
```diff
@@ -293,6 +293,985 @@ hive>
+2018-09-24 22:20:00 INFO  OutputCommitCoordinator$OutputCommitCoordinatorEndpoint:54 - OutputCommitCoordinator stopped!
+2018-09-24 22:20:00 INFO  SparkContext:54 - Successfully stopped SparkContext
+# Run the following spark-submit command to execute the delta-streamer and ingest to stock_ticks_mor dataset in HDFS
+spark-submit --class com.uber.hoodie.utilities.deltastreamer.HoodieDeltaStreamer /var/hoodie/ws/hoodie-utilities/target/hoodie-utilities-0.4.4-SNAPSHOT.jar --source-class com.uber.hoodie.utilities.sources.JsonKafkaSource --source-ordering-field ts  --target-base-path /user/hive/warehouse/stock_ticks_mor --target-table stock_ticks_mor --props /var/demo/config/kafka-source.properties
```
this is now `hoodie-utilities-0.4.4-SNAPSHOT.jar` with the new release. same comment on other 3 occurrances as well.. Can we name the jar differently inside the container without the version? that way it will continue to work? (I understand it takes away knowing what version is being tested.) thoughts?

vinothchandar replied:
Also DeltaStreamer now takes a `--storage-type COPY_ON_WRITE/MERGE_ON_READ` argument which is required.. we can use this and get rid of steps to create the dataset manually via CLI? DeltaStreamer will create the dataset if the basePath does not exist.

vinothchandar replied:
@bvaradar https://github.com/docker/compose/issues/3574 don't think `compose up` will pull in latest images.. I am going remove all containers and retry.. We probably need to have a better support in the script for optionally also doing a `compose pull` before?

vinothchandar replied:
That worked @bvaradar .. One more issue I found was between recreating containers, the path on host machine `/tmp/hadoop*` needs to be blown away.. Can we add a `rm -rf` before the `mkdir -p` in the setup script.. other than that, verified that building and deltastreamer work..

vinothchandar commented on docs/quickstart.md:
```diff
@@ -293,6 +293,985 @@ hive>
+Lets run similar queries against M-O-R dataset. Lets look at both 
+ReadOptimized and Realtime views supported by M-O-R dataset
+
+# Run agains ReadOptimized View. Notice that the latest timestamp is 10:29
```
typo: against

vinothchandar commented on docs/quickstart.md:
```diff
@@ -293,6 +293,985 @@ hive>
+1 row selected (6.326 seconds)
+
+
+# Run agains Realtime View. Notice that the latest timestamp is again 10:29
```
typo: against

vinothchandar commented on docs/quickstart.md:
````diff
@@ -293,6 +293,985 @@ hive>
+running in spark-sql
+
+```
+$SPARK_INSTALL/bin/spark-shell --jars $HUDI_SPARK_BUNDLE --master local[2] --driver-class-path $HADOOP_CONF_DIR --conf spark.sql.hive.convertMetastoreParquet=false --deploy-mode client  --driver-memory 1G --executor-memory 3G --num-executors 1  --packages com.databricks:spark-avro_2.11:4.0.0
````
add `docker exec -it adhoc-1 /bin/bash` ?

vinothchandar commented on docs/quickstart.md:
```diff
@@ -293,6 +293,985 @@ hive>
+
+#### Step 7(b): Run Spark SQL Queries
+
+Running the same queries in Spark-SQl:
```
nit: SQL

vinothchandar commented on docker/demo/config/kafka-source.properties:
```diff
@@ -0,0 +1,30 @@
+hoodie.deltastreamer.schemaprovider.source.schema.file=/var/demo/config/schema.avsc
+hoodie.deltastreamer.schemaprovider.target.schema.file=/var/demo/config/schema.avsc
+# Kafka Source
+#hoodie.deltastreamer.source.kafka.topic=uber_trips
```
nit: remove this line?

vinothchandar commented on docker/hoodie/hadoop/base/docker-maven-plugin-1.1.0.pom:
```diff
@@ -0,0 +1,433 @@
+  <packaging>maven-plugin</packaging>
+  <name>docker-maven-plugin</name>
+  <description>A maven plugin for docker</description>
+  <url>https://github.com/spotify/docker-maven-plugin</url>
```
is the plugin not published anywhere we can pull down from?

vinothchandar commented:
@bvaradar reported two issues with respect to reiniting containers.. otherwise its good for merging. I will go ahead and do that

vinothchandar commented on docker/hoodie/hadoop/hive_base/README.md:
```diff
@@ -0,0 +1,49 @@
+[![Gitter chat](https://badges.gitter.im/gitterHQ/gitter.png)](https://gitter.im/big-data-europe/Lobby)
+
+# docker-hive
```
@bvaradar is this file meant to be checked in?

bvaradar replied:
Missed this. I have removed it now.

vinothchandar commented on docker/setup_demo.sh:
```diff
@@ -0,0 +1,15 @@
+# Create host mount directory and copy
+mkdir -p /tmp/hadoop_name
```
add a protective `rm -rf` here? otherwise NN will start in safe mode since it does not recognize the data

bvaradar replied:
once the docker containers are stopped, the  files are deleted using "rm -rf ".  I did not want to do the deletion before docker compose down so that docker containers can be shutdown more cleanly. Will chat f2f to explain more.

vinothchandar commented:
@bvaradar actually lets do a quick sync before I merge this

bvaradar commented:
@vinothchandar : Changed the setup_demo script to first do docker-compose pull before docker-compose up in order to pull the latest version of docker images. Addressed other comments.
